### PR TITLE
Validation

### PR DESCRIPTION
**Pull request by mw-lb**

### CODE REVIEW

File: kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/DeploymentsValidation.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+object DeploymentsValidation {
+  import TopologyValidation.*
+
+  def validate(topologiesBefore: Topologies, deployments: Deployments): ValidationFailures = {
+    // First validate the deployments itself on its own
+    val deploymentValidation = validateDeployments(deployments)
+
+    // Then pretend we applied the new topologies and the removed ones and...
+    val topologiesAfter = topologiesBefore ++ deployments.deploy.toMapById -- deployments.delete
```
Comment (mmniazi) [inline kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/DeploymentsValidation.scala:20]: minor nitpick `deployments.deploy` is a bit confusing to read, I had to check `Deployments` class to confirm that this is just Seq[Topology]

File: kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/DeploymentsValidation.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  def validateDeployments(deployments: Deployments): ValidationFailures = {
+    // Validating all new topologies on their own
+    val newTopologiesValidation = deployments.deploy.map(validateTopology).combine
+
+    // Validating the deployment itself
+    val duplicateDeployTopologyIds = deployments.deploy.duplicatesBy(_.id)
+    val uniqueDeployTopologiesValidation = validationErrorIf(s"cannot deploy duplicate topologies ${duplicateDeployTopologyIds.map(_.quote).mkString(", ")}")(duplicateDeployTopologyIds.nonEmpty)
+    val duplicateDeleteTopologyIds = deployments.delete.duplicates
+    val uniqueDeleteTopologiesValidation = validationErrorIf(s"cannot delete duplicate topologies ${duplicateDeleteTopologyIds.map(_.quote).mkString(", ")}")(duplicateDeleteTopologyIds.nonEmpty)
```
Comment (mmniazi) [inline kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/DeploymentsValidation.scala:37]:
```suggestion
    val uniqueDeleteTopologiesValidation = validationErrorIf(s"cannot have duplicates in delete topologies list ${duplicateDeleteTopologyIds.map(_.quote).mkString(", ")}")(duplicateDeleteTopologyIds.nonEmpty)
```

File: kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/TopologyValidation.scala [hunk 3]
```diff
@@ -6,30 +6,32 @@
...
-    validateTopologyId(topology.id) + validateNamespace(topology.namespace)
+  def validateTopologyDependencies(@unused topology: Topology, @unused topologies: Topologies): ValidationFailures = {
+    // TODO implement
+    valid
   }
 
-  def validateNamespace(namespace: Namespace): ValidationFailures = {
-    validationErrorIf(s"namespace ${namespace.quote} must not start or end with '.', '_', '_' and can contain only alphanumeric characters and '.', '-', '_'") {
+  private def validateNamespace(namespace: Namespace): ValidationFailures = {
+    validationErrorIf(s"namespace ${namespace.quote} must not start or end with '.' and can contain only alphanumeric characters and '.', '-', '_'") {
```
Comment (mmniazi) [inline kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/TopologyValidation.scala:28]: not sure if I understand this one, we are saying that make sure namespace is not empty and it has nothing after `.` (that's what I got from regex above). But in error message we make it sound like `.` is allowed in namespace

Comment (mmniazi) [reply kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/TopologyValidation.scala:28]: looking at tests, the error message seems correct. Maybe I misunderstood what that regex is doing

**Review by mmniazi — APPROVED** [review]
Looks good, left some cosmetic comments but feel free to ignore them

File: kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/DeploymentsValidation.scala (hunk 1 above)
Comment (mw-lb) [reply kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/DeploymentsValidation.scala:20]: agree, my goal was to make json/yaml more readable/writable. Do you have any other name ideas that look good both in the json/yaml and code? Alternatively we can use different names in json/yaml, but that sounds a bit overkill.

File: kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/TopologyValidation.scala (hunk 3 above)
Comment (mw-lb) [reply kafkakewl-domain/src/main/scala/com/mwam/kafkakewl/domain/validation/TopologyValidation.scala:28]: It expects non-empty namspaces to start with a fragment of letters/numbers/hyphen/underscore, then zero more fragments that start with dot and followed by one ore more letters/numbers/hyphen/underscore.

I think it does the job?

**Pull request merged**